lib/galaxy/tools/error_reports/plugins/gitlab.py

```diff
@@ -0,0 +1,144 @@
+
+        try:
+            import gitlab
+            import requests
+            import os
+
+            session = requests.Session()
+            session.proxies = {
+                'https': os.environ.get('https_proxy'),
+                'http': os.environ.get('http_proxy'),
+            }
+            self.gitlab = gitlab.Gitlab(
+                # Allow running against GL enterprise deployments
+                kwargs.get('gitlab_base_url', 'https://gitlab.com'),
+                private_token=kwargs['gitlab_private_token'],
+                session=session
+            )
+            self.gitlab.auth()
+
+        except ImportError:
+            log.error("Please install python-gitlab to submit bug reports to gitlab")
+            self.gitlab = None
+
+    def submit_report(self, dataset, job, tool, **kwargs):
+        """Submit the error report to GitLab
+        """
+        log.info(self.gitlab)
+
+        if self.gitlab:
+            # Import the necessary libraries
+            import sys
+            import requests
+            if sys.version_info[0] < 3:
+                import urllib as urllib
+                import urlparse as urlparse
+            else:
+                import urllib.parse as urllib
+                urlparse = urllib
+
+            log.info(job.tool_id)
+            log.info(job.tool_version)
+            log.info(tool.tool_shed)
+
+            # Determine the ToolShed
+            ts_url_request = requests.get('http://' + str(tool.tool_shed))
+            ts_url = ts_url_request.url
+            log.info("Determined ToolShed is " + ts_url)
+
+            # Find the repo inside the ToolShed
+            ts_repo_request = requests.get(ts_url + "/api/repositories?tool_ids=" + str(job.tool_id))
+            ts_repo_request_data = ts_repo_request.json()
+            ts_repourl = None
+
+            for changeset, repoinfo in ts_repo_request_data.items():
+                if isinstance(repoinfo, dict):
+                    if 'repository' in repoinfo.keys():
+                        if 'remote_repository_url' in repoinfo['repository'].keys():
+                            ts_repourl = repoinfo['repository']['remote_repository_url']
+
+            log.info("Determine ToolShed Repository URL: " + ts_repourl)
+
+            if ts_repourl:
+                gitlab_projecturl = urlparse.urlparse(ts_repourl).path[1:]
+            else:
+                gitlab_projecturl = "/".join([self.gitlab_default_repo_owner,
+                                              self.gitlab_default_repo_name])
+
+            gitlab_urlencodedpath = urllib.quote_plus(gitlab_projecturl)
+
+            # Make sure we are always logged in, then retrieve the GitLab project
+            self.gitlab.auth()
+            gl_project = self.gitlab.projects.get(gitlab_urlencodedpath)
+
+            # Get a list of all the issues, parse them for de-duplication later
+            log.info(gl_project.issues.list())
+
+            issue_cache = {}
```

Comment:
Have you considered modifying the issue cache to be as part of `__init__`? Then you could have sub-caches per `gitlab_projecturl`, which might speed up error reporting for tools with large bug lists (and help decrease the chance that you hit API ratelimits).

Reply:
Possibly a clean issue cache on __init__ which gets filled on the first bug report for a specific tool would make most sense in this case. I'll add this.